### PR TITLE
weight_quantize/weight_only_linear support Volta Arch

### PR DESCRIPTION
**Pull request by MARD1NO**
### PR types
New features

### PR changes
OPs

### Description
1. when in volta arch, weightonly needs Row major, we do not need to transpose it. 
2. And we do not write rowmajor gemv, so when M=1 and arch=70, it still use Weightonly Gemm. 
3. Since the layout has changed in sm70, weightonly linear grad maynot suit for sm70, here we just force arch==80 when using weightonly linear grad operator. 

Pcard-72603

### CODE REVIEW

File: paddle/phi/infermeta/multiary.cc [hunk 1]
```diff
@@ -3846,7 +3846,13 @@ void WeightOnlyLinearInferMeta(const MetaTensor& x,
                                const MetaTensor& bias,
                                const MetaTensor& weight_scale,
                                const std::string& weight_dtype,
+                               const int32_t arch,
```
Comment (yuanlehome) [inline paddle/phi/infermeta/multiary.cc]: int类型的值传递就不需要const修饰了

File: python/paddle/nn/quant/quantized_linear.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 from paddle.framework import LayerHelper, in_dynamic_mode
 
 
-def weight_quantize(x, algo="weight_only_int8"):
+def weight_quantize(x, algo="weight_only_int8", arch=80):
     """
```
Comment (yuanlehome) [inline python/paddle/nn/quant/quantized_linear.py:40]: 一定要新增这个参数的话，详细解释下
image

Comment (MARD1NO) [reply python/paddle/nn/quant/quantized_linear.py:40]: 已补充

File: paddle/phi/kernels/cpu/weight_quantize_kernel.cc [hunk 3]
```diff
@@ -19,6 +19,9 @@ limitations under the License. */
 #include "paddle/phi/kernels/funcs/blas/blas.h"
 #include "paddle/phi/kernels/funcs/common_shape.h"
 #include "paddle/phi/kernels/impl/weight_quantize_kernel_impl.h"
+#if defined(PADDLE_WITH_CUTLASS)
```
Comment (yuanlehome) [inline paddle/phi/kernels/cpu/weight_quantize_kernel.cc]: weight_quantize是一个cpu kernel吧，依赖cutlass?

File: paddle/phi/kernels/gpu/weight_only_linear_kernel.cu [hunk 4]
```diff
@@ -30,7 +31,26 @@ void WeightOnlyLinearKernel(const Context& dev_ctx,
...
                             const DenseTensor& weight_scale,
                             const std::string& weight_dtype,
+                            const int32_t arch,
                             DenseTensor* out) {
+  int32_t cuda_arch_version;
+#if defined(PADDLE_WITH_CUTLASS)
+  if (arch == 0) {
+    // Note(Zhengzekang): user do not set the arch, we will get SM Arch from
+    // device.
+    cuda_arch_version = getSMVersion();
```
Comment (yuanlehome) [inline paddle/phi/kernels/gpu/weight_only_linear_kernel.cu]: getsmversion的操作放在quantized_linear.py中 api定义里实现更好？

Comment (MARD1NO) [reply paddle/phi/kernels/gpu/weight_only_linear_kernel.cu]: Done

File: paddle/phi/kernels/cpu/weight_quantize_kernel.cc (hunk 3 above)
Comment (MARD1NO) [reply paddle/phi/kernels/cpu/weight_quantize_kernel.cc]: 已删除

File: paddle/phi/api/yaml/ops.yaml [hunk 5]
```diff
@@ -2807,7 +2807,7 @@
   backward: weight_only_linear_grad
 
 - op : weight_quantize
-  args : (Tensor x, str algo="weight_only_int8")
+  args : (Tensor x, str algo = "weight_only_int8", int arch = -1)
```
Comment (yuanlehome) [inline paddle/phi/api/yaml/ops.yaml]: 改为默认是80

File: python/paddle/nn/quant/quantized_linear.py [hunk 6]
```diff
@@ -146,9 +159,24 @@ def weight_only_linear(
             ...    print(out.shape)
             [1, 2, 32]
     """
+    if arch is None:
+        # Get SMVersion from device.
+        cuda_version = version.cuda()
```
Comment (yuanlehome) [inline python/paddle/nn/quant/quantized_linear.py]: 这块代码封装成一个函数给这两个api调用

**Review by yuanlehome — APPROVED** [review]
LGGGTM

**Review by sunzhongkai588 — APPROVED** [review]
LGTM for docs～

**Review by jeff41404 — APPROVED** [review]
`weight_quantize` and `weight_only_linear` have added  default value parameters `arch=None`which belongs to compatibility upgrade, so approve.

**Review by heavengate — APPROVED** [review]

**Review by zyfncg — APPROVED** [review]

**Pull request merged**